core/transaction-pool/src/api.rs

```diff
@@ -86,3 +94,82 @@ impl<T, Block> txpool::ChainApi for FullChainApi<T, Block> where
 		})
 	}
 }
+
+/// The transaction pool logic for light client.
+pub struct LightChainApi<T, F, Block> {
+	client: Arc<T>,
+	fetcher: Arc<F>,
+	_phantom: PhantomData<Block>,
+}
+
+impl<T, F, Block> LightChainApi<T, F, Block> where
+	Block: BlockT,
+	T: HeaderBackend<Block>,
+	F: Fetcher<Block>,
+{
+	/// Create new transaction pool logic.
+	pub fn new(client: Arc<T>, fetcher: Arc<F>) -> Self {
+		LightChainApi {
+			client,
+			fetcher,
+			_phantom: Default::default(),
+		}
+	}
+}
+
+impl<T, F, Block> txpool::ChainApi for LightChainApi<T, F, Block> where
+	Block: BlockT<Hash=H256>,
+	T: HeaderBackend<Block> + 'static,
+	F: Fetcher<Block> + 'static,
+{
+	type Block = Block;
+	type Hash = H256;
+	type Error = error::Error;
+	type ValidationFuture = Box<dyn Future<Output = error::Result<TransactionValidity>> + Send + Unpin>;
+
+	fn validate_transaction(
+		&self,
+		at: &BlockId<Self::Block>,
+		uxt: txpool::ExtrinsicFor<Self>,
+	) -> Self::ValidationFuture {
+		let header_hash = self.client.expect_block_hash_from_id(at);
+		let header_and_hash = header_hash
+			.and_then(|header_hash| self.client.expect_header(BlockId::Hash(header_hash))
+				.map(|header| (header_hash, header)));
+		let (block, header) = match header_and_hash {
+			Ok((header_hash, header)) => (header_hash, header),
+			Err(err) => return Box::new(ready(Err(err.into()))),
+		};
+		let remote_validation_request = self.fetcher.remote_call(RemoteCallRequest {
+			block,
+			header,
+			method: "TaggedTransactionQueue_validate_transaction".into(),
```

Comment:
Same as with other grumble. We really need to derive these strings from the declaration/implementation.